core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java

```diff
@@ -0,0 +1,283 @@
+        }
+    }
+
+    public void testShardAdditionAndRemoval() {
+        MockController controller = new MockController(Settings.builder()
+                .put(IndexingMemoryController.INDEX_BUFFER_SIZE_SETTING, "10mb")
+                .put(IndexingMemoryController.TRANSLOG_BUFFER_SIZE_SETTING, "100kb").build());
+        final ShardId shard1 = new ShardId("test", 1);
+        controller.setTranslog(shard1, randomInt(10), randomInt(10));
+        controller.forceCheck();
+        controller.assertBuffers(shard1, new ByteSizeValue(10, ByteSizeUnit.MB), new ByteSizeValue(64, ByteSizeUnit.KB)); // translog is maxed at 64K
+
+        // add another shard
+        final ShardId shard2 = new ShardId("test", 2);
+        controller.setTranslog(shard2, randomInt(10), randomInt(10));
+        controller.forceCheck();
+        controller.assertBuffers(shard1, new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(50, ByteSizeUnit.KB));
+        controller.assertBuffers(shard2, new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(50, ByteSizeUnit.KB));
+
+        // remove first shard
+        controller.deleteShard(shard1);
+        controller.forceCheck();
+        controller.assertBuffers(shard2, new ByteSizeValue(10, ByteSizeUnit.MB), new ByteSizeValue(64, ByteSizeUnit.KB)); // translog is maxed at 64K
+
+        // remove second shard
+        controller.deleteShard(shard2);
+        controller.forceCheck();
+
+        // add a new one
+        final ShardId shard3 = new ShardId("test", 3);
+        controller.setTranslog(shard3, randomInt(10), randomInt(10));
+        controller.forceCheck();
+        controller.assertBuffers(shard3, new ByteSizeValue(10, ByteSizeUnit.MB), new ByteSizeValue(64, ByteSizeUnit.KB)); // translog is maxed at 64K
+    }
+
+    public void testActiveInactive() {
+        MockController controller = new MockController(Settings.builder()
+                .put(IndexingMemoryController.INDEX_BUFFER_SIZE_SETTING, "10mb")
+                .put(IndexingMemoryController.TRANSLOG_BUFFER_SIZE_SETTING, "100kb")
+                .put(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING, "5s")
+                .build());
+
+        final ShardId shard1 = new ShardId("test", 1);
+        controller.setTranslog(shard1, 0, 0);
+        final ShardId shard2 = new ShardId("test", 2);
+        controller.setTranslog(shard2, 0, 0);
+        controller.forceCheck();
+        controller.assertBuffers(shard1, new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(50, ByteSizeUnit.KB));
+        controller.assertBuffers(shard2, new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(50, ByteSizeUnit.KB));
+
+        // index into both shards, move the clock and see that they are still active
+        controller.setTranslog(shard1, randomInt(2), randomInt(2) + 1);
+        controller.setTranslog(shard2, randomInt(2) + 1, randomInt(2));
+        controller.incrementTimeSec(10);
```

Comment:
This is a bit confusing: the inactive time is 5s, so after 10s you'd expect the shards to become inactive, but it's only because this next `forceCheck` is seeing the new ops in the translog for the first time, that they remain active right?  Maybe add a comment :)